### PR TITLE
Incremental parsing (WIP)

### PR DESCRIPTION
**Pull request by alexsaveliev**
Important:

You must clear `node_modules/typescript` and run `npm install` after merge to make sure you are using forked version of `microsoft/typescript` located at `alexsaveliev/typescript`

Workspace symbols:
 - in incremental mode we are adding files on demand so some may be left not parsed when we are asking for symbol's references. To deal with it, we'll search for files in the virtual FS using `tsconfig.json` rules the same way Typescript does (`src/match-files.ts` is actually a subset of Typescript's `src/compiler/sys.ts` and `src/compiler/core.ts`) to ensure all project files were collected
 - parsing all project files when asking for symbols
 - searching for workspace symbols in the all sub-projects (example: microsoft/vscode)
 - declaring "workspace symbols" provider capability for VSCode
 - caching workspace symbols

References:
 - paralleling references transformation
 - parsing all project files when asking for references

Performance optimization:
 - caching service, host, and program objects
 - using modified version of microsoft/typescript (2.1.x) that exposes 'addFile' method
 - do not pass any files as root ones when making new program, starting with the empty set and adding files on demand

Misc:
 - added support of multiple subprojects denoted by tsconfig.json files
 - ignoring didOpen/didClose requests for now

JS support
 - detecting projects denoted by `jsconfig.json`, forcing `allowJs` option for these projects
 - enabled JS in the default project
 - when no project files were found, using default options to scan for files available

### CODE REVIEW

File: src/match-files.ts [hunk 1]
```diff
@@ -0,0 +1,573 @@
+/// <reference path="../typings/node/node.d.ts"/>
+/// <reference path="../typings/typescript/typescript.d.ts"/>
+
```
Comment (beyang) [inline src/match-files.ts]: Add a comment indicating much of this file was initially copied over from `src/compiler/sys.ts` and `src/compiler/core.ts`.

File: src/project-manager.ts [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+
+import * as ts from 'typescript';
+import { IConnection } from 'vscode-languageserver';
+import * as async from 'async';
+
+import * as FileSystem from './fs';
+import * as util from './util';
+import * as match from './match-files';
+
+export class ProjectManager {
```
Comment (beyang) [inline src/project-manager.ts]: docstring for this

**Comment by beyang** [comment]
A few minor comments but otherwise LGTM.

A few things I noticed when running this:
- When I type a query via the quick-search modal in the UI, this often leads to the network indicator displaying indefinitely. Do you know why this happens?
- Multiple consecutive requests still seem to block one another. For instance, if I type a query, it'll issue a `workspace/symbol` request for every keystroke (if I type slowly enough). When it does this, the last request takes longer to return, presumably because it blocks on serving prior requests first. Any ideas to fix this?

**Comment by sqs** [comment]
Are the changes to the TypeScript repo going to be submitted upstream? They should be. Just checking.

**Comment by sqs** [comment]
Great. I am seeing about a 30% reduction for an initialize+hover operation on Microsoft/vscode. See below.

It seems we can still speed this up a lot by not reading unnecessary `tsconfig.json` files; see the alexsaveliev/master logs below.

## sourcegraph/master

```
$ time curl -sS --data '[{"id":0,"method":"initialize","params":{"rootPath":"git://github.com/Microsoft/vscode?0cf21a86a10b632afd4334444fb26e50a3ddcc40","mode":"typescript"}},{"id":1,"method":"textDocument/hover","params":{"textDocument":{"uri":"git://github.com/Microsoft/vscode?0cf21a86a10b632afd4334444fb26e50a3ddcc40#src/vs/editor/common/core/editorState.ts"},"position":{"line":26,"character":29}}},{"id":2,"method":"shutdown"},{"method":"exit"}]' http://localhost:3080/.api/xlang/textDocument/hover

real	0m13.906s
```

```
langserver-jsts: initialize file:///
langserver-jsts: 1407 found, fs scan complete in 0.911
langserver-jsts: 1407 found, fs scan complete in 0.912
langserver-jsts: files fetched in 1.092
langserver-jsts: hover file:///src/vs/editor/common/core/editorState.ts 26 29 total 13.629 busy 11.619 wait 2.01
langserver-jsts: files fetched in 13.918
```

## alexsaveliev/master (this PR's changes)

```
$ time curl -sS --data '[{"id":0,"method":"initialize","params":{"rootPath":"git://github.com/Microsoft/vscode?0cf21a86a10b632afd4334444fb26e50a3ddcc40","mode":"typescript"}},{"id":1,"method":"textDocument/hover","params":{"textDocument":{"uri":"git://github.com/Microsoft/vscode?0cf21a86a10b632afd4334444fb26e50a3ddcc40#src/vs/editor/common/core/editorState.ts"},"position":{"line":26,"character":29}}},{"id":2,"method":"shutdown"},{"method":"exit"}]' http://localhost:3080/.api/xlang/textDocument/hover

real	0m9.732s
```

```
langserver-jsts: initialize file:///
langserver-jsts: 1407 found, fs scan complete in 0.971
langserver-jsts: 1407 found, fs scan complete in 0.976
langserver-jsts: files fetched in 0.978
langserver-jsts: Added project jsconfig.json
langserver-jsts: Added project build/tsconfig.json
langserver-jsts: Added project src/tsconfig.json
langserver-jsts: Added project extensions/extension-editing/tsconfig.json
langserver-jsts: Added project extensions/configuration-editing/tsconfig.json
langserver-jsts: Added project extensions/python/tsconfig.json
langserver-jsts: Added project extensions/javascript/tsconfig.json
langserver-jsts: Added project extensions/vscode-colorize-tests/tsconfig.json
langserver-jsts: Added project extensions/markdown/tsconfig.json
langserver-jsts: Added project extensions/php/tsconfig.json
langserver-jsts: Added project extensions/typescript/tsconfig.json
langserver-jsts: Added project extensions/vscode-api-tests/tsconfig.json
langserver-jsts: Added project extensions/css/client/tsconfig.json
langserver-jsts: Added project extensions/css/server/tsconfig.json
langserver-jsts: Added project extensions/html/client/tsconfig.json
langserver-jsts: Added project extensions/html/server/tsconfig.json
langserver-jsts: Added project extensions/json/client/tsconfig.json
langserver-jsts: Added project extensions/json/server/tsconfig.json
langserver-jsts: Added project extensions/typescript/test/colorize-fixtures/tsconfig.json
langserver-jsts: hover file:///src/vs/editor/common/core/editorState.ts 26 29 total 9.346 busy 0.541 wait 8.805
```

**Comment by sqs** [comment]
@alexsaveliev It seems we can still speed this up a lot by not reading unnecessary `tsconfig.json` files (see the logs above with `Added project ...`; all but 1 is unnecessary). If I make it ONLY read the nearest ancestor `[tj]sconfig.json` file, the same `curl` command as above takes 3.6s (way better than even 9.7s).

This diff is what I used to achieve 3.6s on Microsoft/vscode. Obviously it only applies to that project, but the real implementation (only read the nearest ancestor `[tj]sconfig.json` file for hover/definition) should be fairly simple to do.

```diff
diff --git a/src/project-manager.ts b/src/project-manager.ts
index 460e80c..f51fb5b 100644
--- a/src/project-manager.ts
+++ b/src/project-manager.ts
@@ -263,6 +263,7 @@ export class ProjectManager {
         let tasks = [];
         const self = this;
         Object.keys(this.localFs.entries).forEach(function(k) {
+                       if (k !== "src/tsconfig.json") return;
             if (!/(^|\/)[tj]sconfig\.json$/.test(k)) {
                 return;
             }
```

**Pull request merged**